README.md
Respository containing all the tools used for the QoE meets OpenFlow work. Also explains how far we got with experimentation.

Topology & Experiment
--------

<img src=qtest/2tierdigram.png width=600  />

Mu has a diagram of the topology (above), but here is a brief description of it:

There are 5 households, each consisting of between 3 and 6 clients. The clients in a household are connected together by a single OpenFlow switch. These households are the second tier (and served by the ```resourceallocationcaller.py``` functions) This OpenFlow switch is also the gateway out of the local network. These switches are then connected to an arregation switch, which forms the first tier (represented in the ```firsttiercaller.py```). This switch is then connected to another switch, which has directly attached to it the background server and the foreground server.

In reality, each of these hosts and servers are virtual machines sat on the OpenStack testbed. In order to guarantee which port they exit from, they are each connected to a virtual switch with OpenStack. This switch tags them with a specific VLAN. When the traffic exits the OpenStack deployment towards the first physical switch, the VLAN tagging is used to determine which physical port the packets emerged from. 

These physical ports are then connected to a second physical switch. This switch is concertinaed into a number of virtual switches, each acting as a separate entity. The wiring is done to replicate the above topology.

qTest
-----
This is glues together the various elements. Confusingly, it includes both my integration code *and* Mu's QoE code, and can be found in the ```\qtest``` folder. The code to run is the ```\qtest\integration.py``` script. This will do the talking between the OpenFlow controller and Mu's QoE code. Try ```python integration.py -h``` to check out the parameters you can pass. To install the required packages, run ```pip install -r requirements.txt```; that should install everything. 

I've also included the config file used in the experiments (```\qtest\config.json```). It's basically a tree structure, describing the nodes (hosts, servers and switches) and the connections between them (the edges). The config is fairly self explanatory, but is rather tedious to build. 

I've also documented the ```\qtest\integration.py``` code pretty well, so it should be relatively easy to modify it (which you will undoubtedly have to). There is some hardcoded variables at the start (the household mappings, for example). These might need to be changed if you scale things up or name things differently.

OpenFlow Bandwidth
------------------

The modified Ryu controller used in our experiments can be found in the ```/openflow_bandwidth``` folder. It is almost identical to the version originally developed by Jamie and Lyndon, and further developed by Nic. The latest version can be found [here](https://github.com/hdb3/openflow_bandwidth). It might be worth combining the changes in a pull request so that we can have a definitive version. I would imagine we will want to expand the interface soon anyway. It worth noting that documentation in Nic's version is old (*hint hint*), and doesn't reflect what is actually implemented.

The version I used has minor changes to the parameters in the *enforce_services* RPC call. It now accepts a list of destinations, rather than single item. This enables the required higher tier functionality, whereby a flow can be limited from the server to a group of destinations (belonging to a single household). This list produces a number of rules that point to the same shared meter.

Scootplayer
-----------
An unmodified version of Scootplayer was used. This can be found [here](https://github.com/broadbent/scootplayer). Ask Mu for the dataset. I'll see if we still have the various MPDs, but they are easy to create and modify.

Tools
-----

There is a number of additional scripts I used in the experimentation. For example, the ```/tools/reboot.sh``` script is used to reboot all the nodes via the nova interface. It could be used again (assuming we naming the hosts and servers the same). This naming shouldn't be an issue if we use the ```/tools/specmuall.py``` script in conjunction with Nic's [ministack tools](https://github.com/hdb3/ministack). This will recreate the exact same configuration of VMs and networks that we used in the experiment. The only think missing is the configuration of the HP switches and the VLAN trickery used to ensure VMs appear on particular ports on each of the *virtual* switches present on the physical HP switch.

There is two bash scripts used during the experiments. These were run simultaneously on each of the VMs using [cluster SSH](http://bit.ly/1dKcMNz). The ```/tools/pre.sh``` script was run before experiments started and basically ensured that there was connectivity between each of the hosts and the server. It prints a nice green message if there is, a big red ```[FAIL]``` if not. This helped debug connectivity issued before even before we started. The ```/tools/go.sh``` was run in a similar manner ()using cluster SSH) to start the tests. The one included is for the background traffic generation using *wget*; evidently, it will different on the hosts playing video (I forgot to retrieve that one, but it basically started Scootplayer pointing at the server and playing the desired version of Big Buck Bunny). I just called it the same so that you could simply hit ```$ ./go.sh``` and start the damned thing!

Samples
-------

I included  a sample output from the integration code in ```/samples/debug.log```. This is basically so that you know what the output will be like. This is also what Mu needs to plot. I also included the ```/samples/history.txt``` which is a dump of the bash history. It should give you an idea of what I was doing to run the experiment (and test things earlier on).

Current Limitations
-------------------
At the moment, the API does not allow the mixing of flow-based and port-based monitoring. This is because a packet will always match on the flow-based rule before the port-based rule (it has more detail, and is thus more specific; OpenFlow will always match on that first). It would be better if the flow-rule then matched on a port-rule, thus combining the two.

There also appears to be an arbitary conversion factor between some of the calls. In the current implementation, I tuned this with some trial and error, so they may not be exactly right. This needs some time to confirm the responses from the API (and the underlying OpenFlow) are consistent. For example, the measurement from the switch does not seem to be the in the same format as the rate-limiting threshold. However, I did confirm that the results (at least those derviced from *wget* and *iperf*) where what I expected. Further experiments need to be done to confirm the result from Scootplayer are what we expect (and no, #inb4scootplayerisbroken).



samples/history.txt
   13  sudo pip install ryu
   14  ls -l
   15  git clone https://github.com/hdb3/openflow_bandwidth.git
   16  sudo apt-get install git
   17  git clone https://github.com/hdb3/openflow_bandwidth.git
   18  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
   19  pip install stevedore
   20  sudo pip install stevedore
   21  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
   22  pip install greenlet
   23  sudo pip install greenlet
   24  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
   25  sudo pip install pysjsonrpc
   26  sudo sudo pip install pysjsonrpc
   27  sudo sudo pip install pip install python-jsonrp
   28  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
   29  sudo pip install python-jsonrpc
   30  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
   31  nano openflow_bandwidth/bandwidth_control_simple_switch_13.py -c
   32  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
   33  ping 192.168.1.201
   34  ping 192.168.1.235
   35  apt-get install clusterssh
   36  sudo apt-get install clusterssh
   37  sudo nano /etc/clusters
   38  clusterssh all
   39  ls -l
   40  nano .ssh/id_rsa.pem
   41  ssh 192.168.1.235
   42  ssh ubuntu@192.168.1.235
   43  ssh ubuntu@192.168.1.201
   44  ls -l .ssh/
   45  ssh ubuntu@192.168.1.201 -v
   46  rm -rf .ssh/known_hosts 
   47  mv .ssh/id_rsa.pem .ssh/id_rsa
   48  ssh ubuntu@192.168.1.201 -v
   49  chmod 0644 .ssh/id_rsa 
   50  ssh ubuntu@192.168.1.201 -v
   51  chmod 700 .ssh/id_rsa 
   52  ssh ubuntu@192.168.1.201 -v
   53  clusterssh all
   54  exit
   55  clusterssh all
   56  exit
   57  clusterssh all
   58  exit
   59  clusterssh all
   60  exit
   61  clusterssh all
   62  exit
   63  clusterssh all
   64  exit
   65  clusterssh all
   66  exit
   67  clusterssh all
   68  exit
   69  clusterssh all
   70  exit
   71  clusterssh all
   72  exit
   73  clusterssh all
   74  exit
   75  clusterssh all
   76  exit
   77  clusterssh all
   78  exit
   79  ssh ubuntu@192.168.1.235 -v
   80  ping 192.168.1.235
   81  ssh ubuntu@192.168.1.235 -v
   82  clusterssh all
   83  clusterssh hosts360
   84  clusterssh hosts720
   85  clusterssh hosts1080
   86  exit
   87  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
   88  killall python
   89  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
   90  clusterssh hosts1080
   91  clusterssh hosts720
   92  clusterssh all
   93  clusterssh all -h
   94  clusterssh all -o "-x -o ConnectTimeout=10" -a './pre.sh'
   95  ssh 192.168.1.234
   96  ssh 192.168.1.235
   97  ssh 192.168.1.235 -h
   98  clusterssh all -o "-x -o ConnectTimeout=10" -a './pre.sh'
   99  ssh 192.168.1.234
  100  ssh 192.168.1.235 
  101  ssh 192.168.1.235  -v
  102  clusterssh all -o "-x -o ConnectTimeout=10"
  103  clusterssh all -o "-x -o ConnectTimeout=30"
  104  clusterssh hosts720
  105  clusterssh all -o "-x -o ConnectTimeout=30"
  106  exit
  107  clusterssh all -o "-x -o ConnectTimeout=30"
  108  clusterssh hosts720
  109  exit
  110  clusterssh hosts720
  111  rm -rf .ssh/known_hosts 
  112  clusterssh hosts720
  113  exit
  114  clusterssh hosts720
  115  clusterssh hosts1080
  116  clusterssh hosts360
  117  exit
  118  clusterssh hosts360
  119  clusterssh foregroundHosts
  120  cat /etc/clusters 
  121  clusterssh hostsForeground
  122  exit
  123  clusterssh hostsForeground
  124  clusterssh hostsBackground
  125  exit
  126  clusterssh hostsBackground
  127  exot
  128  exit
  129  clusterssh hostsBackground
  130  exit
  131  clusterssh all
  132  clusterssh hostsBackground
  133  exit
  134  ryu-manager openflow_bandwidth/bandwidth_control_simple_switch_13.py 
  135  clusterssh hostsBackground
  136  cd qtest
  137  ls -l
  138  cd qtest/
  139  ls -l
  140  nano resourceallocationcaller.py
  141  nano firsttiercaller.py
  142  nano resourceallocationcaller.py
  143  cd ..
  144  python integration -i 30
  145  python integration.py -i 30
  146  cat debug.log 
  147  nano qtest/resourceallocationcaller.py
  148  python integration.py -i 30
  149  ls -l
  150  nano qtest/resourceallocationcaller.py
  151  python integration.py -i 30
  152  nano qtest/resourceallocationcaller.py
  153  python integration.py -i 30
  154  nano qtest/resourceallocationcaller.py
  155  python integration.py -i 30
  156  ls -l
  157  cd qtest/
  158  ls -l
  159  cd ..
  160  nano qtest/resourceallocationcaller.py
  161  python integration.py -i 30
  162  ls -l
  163  nano qtest/resourceallocationcaller.py
  164  python integration.py -i 30
  165  ls -l
  166  nano qtest/resourceallocationcaller.py
  167  python integration.py -i 30
  168  ls -l
  169  nano qtest/resourceallocationcaller.py
  170  python integration.py -i 30
  171  ls -l
  172  cat debug.log 
  173  rm -rf debug.log 
  174  python integration.py -i 30
  175  cat debug.log 
  176  nano qtest/resourceallocationcaller.py
  177  rm -rf debug.log 
  178  python integration.py -i 30
  179  cat debug.log 
  180  rm -rf debug.log 
  181  nano qtest/resourceallocationcaller.py
  182  python integration.py -i 30
  183  cd ..
  184  ls -l
  185  nano pre.sh 
  186  rm -rf pre.sh 
  187  nano pre.sh
  188  chmod +x pre.sh 
  189  ./pre.sh 
  190  nano pre.sh
  191  ./pre.sh 
  192  nano pre.sh
  193  ./pre.sh 
  194  tput
  195  tput setaf1; echo test
  196  tput setaf 1; echo test
  197  tput setaf 1; echo test; tput setaf 0
  198  tput setaf 1; echo test; tput sgr0 0
  199  nano pre.sh 
  200  ./pre.sh 
  201  nano pre.sh 
  202  ssh 192.168.1.235
  203  iperf -h
  204  sudo apt-get install iperf
  205  iperf -h
  206  cd qtest/
  207  python integration.py -i 30
  208  nano qtest/resourceallocationcaller.py
  209  sudo nano /etc/clusters 
  210  cd ..
  211  echo $((RANDOM%200+100))
  212  echo $((RANDOM%400000+2000000)
  213  iperf -c 192.168.1.224 -u 2001814
  214  iperf -c 192.168.1.224 -b 2001814

qtest/integration.py
import qtest.firsttiercaller as firsttiercaller
import qtest.resourceallocationcaller as secondtiercaller
import pyjsonrpc
import json
import threading
import random
import logging
import calendar
import time

from pygraph.classes.digraph import digraph
from optparse import OptionParser

class Integration(object):

    _testing = False #Change to true to include randomly generated test data instead of a real response
    first = []
    second = []
    limits = []
    tiers = ["first", "second"]
    _switch_port_results= {}
    _first_tier_result_mapping = ['A', 'B', 'C', 'D', 'E']

    def __init__(self, config, poll, threshold, capacity, host, port):
        self._parse_graph(self._load_graph(config))
        self._controller = self.Controller(host, port, self._testing)
        self._experience = self.Experience()
        #self._initialise_switch_ports()
        self._threshold = threshold
        self._poll(float(poll))

    def _load_graph(self, path):
        '''Load the JSON from a file.'''
        with open(path) as json_data:
            graph = json.load(json_data)
            json_data.close()
        return graph

    def _parse_graph(self, graph):
        '''Parse the JSON into a graph representation.'''
        self._graph = digraph()
        self._parse_nodes(graph)
        self._parse_edges(graph)

    def _parse_nodes(self, tree):
        '''Parse the nodes from JSON file into objects.'''
        for _type, node in tree["nodes"].iteritems():
            for _id, attrs in node.iteritems():
                attrs["type"] = _type
                self._graph.add_node(node=_id, attrs=attrs.items())
                try:
                    tier = getattr(self, attrs["tier"])
                    tier.append(_id)
                except KeyError:
                    pass

    def _parse_edges(self, tree):
        '''Parse the edge of the graph.'''
        for _id, attrs in tree["edges"].iteritems():
            self._graph.add_edge(edge=tuple(attrs["items"]), wt=1, label=_id, attrs=attrs.items())
            if "limit" in attrs.keys():
                self.limits.append(tuple(attrs["items"]))

    # def _initialise_switch_ports(self):
    #     '''For each of the edges, clear the current meters.'''
    #     for edge in self.limits:
    #         attrs = dict(self._graph.edge_attributes(edge))
    #         switch = self._get_field_from_node(attrs["items"][0], "dpid")
    #         self._controller.call(method="enforce_port_outbound", params=[switch, attrs["port"], attrs["limit"]])

    def _poll(self, poll):
        '''At a regular interval, poll each tier for their metrics.'''
        threading.Timer(poll, self._poll, [poll]).start()
        for tier in self.tiers:
            self._fetch_stats(tier)

    def _get_field_from_node(self, node, field):
        '''For a given node, retrieve a specific field value.'''
        try:
            return dict(self._graph.node_attributes(node))[str(field)]
        except KeyError:
            return {}

    def _get_field_from_edge(self, edge, field):
        '''For a given field, retrieve a specific field value.'''
        try:
            return dict(self._graph.edge_attributes(edge))[str(field)]
        except KeyError:
            return {}

    def _fetch_stats(self, tier):
        '''
        For a set of given nodes, retrieve the statistics from the OpenFlow controller.

        Check if a change has occured. If so, recalculate the QoE fairness.
        '''
        nodes = getattr(self, tier)
        for node in nodes:
            switch = self._get_field_from_node(node, "dpid")
            result = self._controller.call(method="report_switch_ports", params=[False, False, switch])
	    if self._compare_switch_ports(tier, switch, result):
                 self._recalculate(tier, switch)

    def _compare_switch_ports(self, tier, switch, result):
        '''Check if a statistic has changed drastically in the result.'''
        if switch not in self._switch_port_results.keys():
            self._switch_port_results[switch] = {}
        for port, throughput in result.iteritems():
            if port not in self._switch_port_results[switch].keys():
                self._switch_port_results[switch][port] = throughput[2]
                continue
            else:
                current = throughput[1]
                previous = self._switch_port_results[switch][port]
            self._switch_port_results[switch][port] = throughput[2]
            return self._calculate_difference(current, previous)

    def _calculate_difference(self, current, previous):
        '''Calculate the difference between the current and previous values.'''
        difference = current - previous #download B/s
        if abs(difference) >= self._threshold:
            return True

    def _recalculate(self, tier, switch):
        '''Call the correct recalculation function given the tier.'''
        getattr(self, '_recalculate_' + tier + '_tier')(switch)

    def _recalculate_first_tier(self, _):
        '''Recalculate the metrics for the first tier.'''
        totalbw, households = self._fetch_first_tier_stats()
        result = self._experience.first(totalbw=totalbw, households=households)
        switch = self._get_field_from_node(self.first[0], "dpid")
        result = self._fix_household_result(switch, result)
        self._effect_first_tier_change(switch, result)

    def _recalculate_second_tier(self, switch):
        '''Recalculate the metrics for the second tier.'''
        totalbw, clients, _ = self._fetch_second_tier_stats(switch)
        result = self._experience.second(totalbw=totalbw, clients=clients)
        self._effect_second_tier_change(switch, result)
        self._update_forgiveness_effect(result)

    def _update_forgiveness_effect(self, result):
        '''If a change has been made to the meters, update the forgiveness effect in the QoE function.'''
        timestamp = calendar.timegm(time.gmtime())
        for client_id, allocation in result.iteritems():
            self._experience.forgiveness_effect(client=client_id, timestamp=timestamp, bitrate=allocation[3])

    def _effect_second_tier_change(self, switch, result):
        '''Install a service meter in the second tier.'''
        for id_, allocation in result.iteritems():
            limit = allocation[3]
            limit = self._convert_kilobits_to_bits(limit)
	    src = '192.168.1.235'
	    ip = [self._get_field_from_node(id_, 'ip')]
	    self._controller.call(method="enforce_service", params=[str(switch), src, ip, limit])

    def _effect_first_tier_change(self, switch, result):
        '''Install a service meter in the first tier.'''
        for household in result:
            neighbor = self._find_node_from_label("household", household["household_id"])
        src = '192.168.1.235'
        limit = household["limit"]
        limit = self._convert_kilobits_to_bits(limit)
        dsts = self._fetch_ips_from_household(neighbor)
        self._controller.call(method="enforce_service", params=[str(switch), src, dsts, limit])

    def _fetch_ips_from_household(self, node):
        '''Fetch a the IPs of the hosts within a given household.'''
        ips = []
        neighbors = self._graph.neighbors(node)
        for neighbor in neighbors:
        	ip = self._get_field_from_node(neighbor, 'ip')
        	if ip:
        		ips.append(ip)
        return ips

    def _fetch_first_tier_stats(self):
        '''Fetch the statistics from each of the switches attached to the aggregation (first tier) switch.'''
        households = []
        totalbw = 0
        background = 0
        # node = self.first[0]
        # neighbors = self._graph.neighbors(node)
        for second in self.second:
            household_available, _, household_background = self._fetch_second_tier_stats(second, dpid=False)
            background += household_background
            id_ = self._get_field_from_node(second, "household")
            households.append((id_, household_available))
        # for neighbor in list(set(neighbors)-set(self.second)): #probably only one
        #     port = self._get_field_from_edge((node, neighbor), "port")
        #     switch = self._get_field_from_node(node, "dpid")
        #     totalbw += self._controller.call(method="report_port", params=[False, True, switch, port])[3] #Rx - link max
        totalbw = 100000 #TODO: Hard-coded according to experimental parameters as passive measurement not exercising full link capacity
        totalbw = self._convert_bits_to_kilobits(totalbw)
        totalbw = totalbw - background
        return (totalbw, households)

    def _fetch_second_tier_stats(self, switch, dpid=True):
        '''Fetch the statistics from each of the hosts attached to a switch.'''
        clients = []
        totalbw = 0
        if dpid:
            node = self._find_node_from_label("dpid", switch)
        else:
            node = switch
        neighbors = self._classify_neighbors(node)
        for foreground in neighbors["foreground"]:
            clients.append(self._fetch_foreground(node, foreground))
        totalbw, background = self._fetch_switch(node, neighbors["switch"][0], neighbors["background"])
        return (totalbw, clients, background)

    def _classify_neighbors(self, node):
        '''Classify the neighbours into foreground and background hosts.'''
        neighbors = {}
        for neighbor in self._graph.neighbors(node):
            _type = self._get_field_from_node(neighbor, "type")
            if not neighbors.has_key(_type):
                neighbors[_type] = []
            neighbors[_type].append(neighbor)
        return neighbors

    def _fetch_foreground(self, node, neighbor):
        '''
        Fetch the foreground traffic for a node.

        Assumes no background traffic.
        '''
        port = self._get_field_from_edge((node, neighbor), "port")
    	switch = self._get_field_from_node(node, "dpid")
    	result = self._controller.call(method="report_port", params=[False, True, switch, port])
    	available_bandwidth = result[2] #Tx - link max - no background to remove
    	available_bandwidth = self._convert_bits_to_kilobits(available_bandwidth)
    	resolution = self._get_field_from_node(neighbor, "resolution")
        return ((neighbor, available_bandwidth, resolution))

    def _convert_bits_to_kilobits(self, value):
        '''Convert bits to kilobits.'''
        return value/1000 #TODO: This is a conversion that works, but not correct for bits to kilobits.

    def _convert_kilobits_to_bits(self, value):
        '''Convert kilobits to bits.'''
        return value * 2 #TODO: This is a conversion that works, but not correct for kilobits to bits.

    def _fetch_switch(self, node, neighbor, background):
        '''Fetch the background traffic and available bandwidth for a given switch.'''
        background_traffic = 0
        switch = self._get_field_from_node(node, "dpid")
        port = self._get_field_from_edge((node, neighbor), "port")
        #max_bandwidth = self._controller.call(method="report_port", params=[False, True, switch, port])[3] #Rx - link max
        max_bandwidth = 20000 #TODO: Hard-coded according to experimental parameters
        for client in background:
            port = self._get_field_from_edge((node, client), "port")
        background_traffic += self._controller.call(method="report_port", params=[False, False, switch, port])[2] #Tx - current background
        background_traffic = self._convert_bits_to_kilobits(background_traffic)
        available_bandwidth = max_bandwidth - background_traffic
        try:
            assert available_bandwidth >= 0
        except AssertionError:
            print available_bandwidth, max_bandwidth, background_traffic
        return available_bandwidth, background_traffic

    def _find_node_from_label(self, field, value):
        '''Find a node from a given label.'''
        for node in self._graph.nodes():
            if self._get_field_from_node(node, field) == value:
                return node

    def _fix_household_result(self, switch, result):
        '''Map index in result to household ID. Fixed mapping (see object variables).'''
        limits = []
        for index, limit in enumerate(result):
            household_id = self._first_tier_result_mapping[index]
            neighbor = self._find_node_from_label("household", household_id)
            node = self._find_node_from_label("dpid", switch)
            port = self._get_field_from_edge((node, neighbor), "port")
            limits.append({'household_id' : household_id, 'port' : port, 'limit' : limit})
        return limits

    class Controller(object):
        '''Represents all communication with the controller.'''

        def __init__(self, host, port, testing):
            '''Connect to the controller.'''
            self._testing = testing
            self._client = pyjsonrpc.HttpClient(url = "http://" + host + ":" + port + "/jsonrpc")

        def call(self, **kwargs):
            '''Make a call to the controller.'''
            result = None
            logging.debug('[controller][call]: %s', kwargs)
            if self._testing:
                if kwargs['method'] == 'report_switch_ports':
                    result = { "1":self._generate_random_bandwidth(4),
                    "2":self._generate_random_bandwidth(4),
                    "3":self._generate_random_bandwidth(4),
                    "4":self._generate_random_bandwidth(4),
                    "5":self._generate_random_bandwidth(4),
                    "6":self._generate_random_bandwidth(4),
                    "7":self._generate_random_bandwidth(4)}
                elif kwargs['method'] == "enforce_service":
                    result = random.randint(53, 200)
                elif kwargs['method'] == "report_port":
                    result = self._generate_random_bandwidth(4)
            else:
                try:
                    result = self._client.call(kwargs['method'], *kwargs['params'])
                except pyjsonrpc.rpcerror.InternalError as e:
                    print e, kwargs
                #result = self._client.notify(kwargs['method'], *kwargs['params'])
                #result = None
                logging.debug('[controller][result]: %s', result)
            return result

        def _generate_random_bandwidth(self, length):
            '''Generate a random bandwidth for testing purposes.'''
            _max = 20000000
            _min = 300000
            bandwidth = []
            for _ in range(length):
                bandwidth.append(random.randint(_min, _max))
            return bandwidth

    class Experience(object):
        '''Represents all communication with the QoE code.'''

        def __init__(self):
            '''Initialise the two tier objects.'''
            self.first_tier = firsttiercaller.FirstTier()
            self.second_tier = secondtiercaller.SecondTier()

        def first(self, **kwargs):
            '''Make a call to the first tier code.'''
            logging.debug('[experience][first][call]: %s', kwargs)
            result = self.first_tier.call(**kwargs)
            logging.debug('[experience][first][result]: %s', result)
            return result

        def second(self, **kwargs):
            '''Make a call to the second tier code.'''
            logging.debug('[experience][second]: %s', kwargs)
            result = self.second_tier.call(**kwargs)
            logging.debug('[experience][second][result]: %s', result)
            return result

        def forgiveness_effect(self, **kwargs):
            '''Update the forgiveness effect. Only present in the second tier.'''
            logging.debug('[experience][forgiveness]: %s', kwargs)
            self.second_tier.set_session_index(**kwargs)

if __name__ == '__main__':
    parser = OptionParser()
    parser.add_option("-i", "--interval", dest="interval", help="controller polling interval, measured in seconds", default=5.0)
    parser.add_option("-t", "--threshold", dest="threshold", help="change threshold at which to trigger a recalculation, measured in bytes", default=1000)
    parser.add_option("-c", "--capacity", dest="capacity", help="maximum capacity to initialise meter to, measured in bytes", default=1000000000)
    parser.add_option("-n", "--hostname", dest="host", help="controller hostname", default="localhost")
    parser.add_option("-p", "--port", dest="port", help="controller interface port", default=4000)
    (options, args) = parser.parse_args()
    logging.basicConfig(filename='debug.log',level=logging.DEBUG, format='[%(asctime)s:%(levelname)s]%(message)s')
    integration = Integration("config.json", float(options.interval), int(options.threshold), int(options.capacity), options.host, str(options.port))
